src/words/TokTok.jl

```diff
@@ -0,0 +1,245 @@
+Handles the following rules using TokenBuffer API:  
+FINAL_PERIOD_1
+FINAL_PERIOD_2
+"""
+function handle_final_periods(ts::TokenBuffer)
+    flag = length(ts.input)
+    # handles FINAL_PERIOD_1 = r"(?<!\.)\.$"
+    if length(ts.input) >= 2 && ts.input[end] == "." && ts.input[end-1] != "."
+        flush!(ts, ".")
+        flag -= 1
+    end
+
+    # handles FINAL_PERIOD_2 = r"(?<!\.)\.\s*(["'’»›”]) *$"
+    if ts.input[end] in string.(Tuple("“”‘’›"))
+        i = flag - 1
+        while ts.input[i] in RSTRIP[1]
+            i -= 1
+        end
+        if ts.input[i] == "."
+            flag = i - 1
+            flush!(ts, ts.input[i:end])
+        end
+    end
+    return flag
+end
+
+"""
+    url_handler1(::TokenBuffer)
+Handles this rule from python using TokenBuffer API:  
+URL_FOE_1 = re.compile(r':(?!//)'), r' : '  
+"""
+function url_handler1(ts::TokenBuffer)
+    str = ":"
+    pattern = "//"
+    if ts.idx + length(pattern) <= length(ts.input) && ts.input[ts.idx] == str && 
+       ts.input[ts.idx + 1 : ts.idx + length(pattern)] != pattern
+        flush!(ts, str)
+        return true
+    elseif ts.idx + length(pattern) <= length(ts.input) && ts.input[ts.idx] == str
+        for i in 1 : length(pattern) + 1
+            push!(ts.buffer, ts.input[ts.idx])
+            ts.idx += 1
+        end
+        return true
+    end
+    return false
+end
+
+"""
+    url_handler2(::TokenBuffer)
+Handles this rule from python using TokenBuffer API:  
+URL_FOE_2 = re.compile(r'\\?(?!\\S)'), r' ? '  
+"""
+function url_handler2(ts::TokenBuffer)
+    str = "?"
+    if ts.idx + 1 <= length(ts.input) && ts.input[ts.idx] == str && isspace(ts.input[ts.idx + 1])
+        flush!(ts, str)
+        return true
+    elseif ts.idx + 1 <= length(ts.input) && ts.input[ts.idx] == str
+        for i in 1 : 2
+            push!(ts.buffer, ts.input[ts.idx])
+            ts.idx += 1
+        end
+        return true
+    end
+    return false
+end
+
+"""
+    url_handler3(::TokenBuffer)
+Handles this rule from python using TokenBuffer API:  
+URL_FOE_3 = re.compile(r'(://)[\\S+\\.\\S+/\\S+][/]'), ' / '
+"""
+function url_handler3(ts::TokenBuffer)
+    str = "://"
+    pattern = [":", "/", "."]
+    if lookahead(ts, str) && ts.idx + 4 <= length(ts.input)
+        if (ts.input[ts.idx + length(str)] in pattern || isspace(ts.input[ts.idx + length(str)])) && 
+            ts.input[ts.idx + length(str) + 1] == "/"
+            flush!(ts, "/")
+            ts.idx += 4
+            return true
+        else
+            for i in 1 : 5
+                push!(ts.buffer, ts.input[ts.idx])
+                ts.idx += 1
+            end
+            return true
+        end
+    end
+    return false
+end
+
+"""
+    url_handler4(::TokenBuffer)
+Handles this rule from python using TokenBuffer API:  
+URL_FOE_4 = re.compile(r' /'), r' / '
+"""
+function url_handler4(ts::TokenBuffer)
+    if lookahead(ts, " /")
+        flush!(ts, "/")
+        ts.idx += 1
```

Comment:
is this right, or should it be `ts.idx += 2`?

Test case hitting this would be useful

Reply:
Yeah, fixed it. Also added the tests covering this.